### DISCUSSION TITLE
Is it possible to script "Source-Set settings" to reference an external file?

### DISCUSSION THREAD

**Discussion by Dexter092** [WarmUpTill/SceneSwitcher#731] [Q&A] [answered]
Greetings!

I must preface **I am no programmer** of sort and all I do is fiddle around stuff until stuff eventually works.
Hence why I love the macros of this plugin as it's very intuitive! :D 

I now unfortunately face an issue that requires better knowledge than what I currently have.

My problem:
I want to make a looping video source (an animation I did) change speed based on an external value, taken from a text file that is streaming such dynamically (the BPM of a song taken from spotify.)

I can set the condition properly. "If file content change, then". That works flawlessly. I see that "source" menu allows to set settings.
Unfortunately, it's a string of command I have no idea what to "tell" to outsource a number to add to a basic math formula (nor I know if what I'm saying can even be done with this function):

{
    "local_file": "PathOfFile.mp4",
    "looping": true,
    "speed_percent": 100
}

I would like to add to the "speed_percent" value something like "(X/120)*100" and then round the number to closest full digit, where X is an outsourced dynamic value from a different file (120 is the standard visual BPM my video is running at).
Can what I'm asking be done with this function? Is there a better function for this?

Thank you kindly for your time.

**Comment by WarmUpTill** [comment, marked as answer]
What you are trying to achieve should be possible by using [variables](https://github.com/WarmUpTill/SceneSwitcher/wiki/Variables).

First you need to set up a macro that assigns the contents of the file of interest to a given variable.
This can be done using a macro similar to this:

![Var1](https://user-images.githubusercontent.com/19472752/235223101-3b6a5078-9da2-4425-ac11-f9ffead0f160.PNG)

Next you can simply reference this variable in your macro that controls the settings of the given source by using this pattern:
`${variable name here}`

So to follow the example from above a macro, that controls the speed of a media source based on the contents of a given file, could look like this:

![Var2](https://user-images.githubusercontent.com/19472752/235223712-7c60971f-8d5b-4c46-818c-7407d6e4fa46.PNG)

I hope that helps! :)
Let me know if you have any questions!

**Comment by Dexter092** [comment]
Greetings! Your explanation worked very well and it partially performs what I want! Thank you so much for this. ^^

By partially I mean two issues:
- I still need a method to apply my basic math formula to convert the change of BPM into a percentage number, to also round the number to the lower value.

- What you made me set works very well, though the conditions seems to not trigger automatically without an additional input for some reason. By changing the value in the file, nothing was apparently happening in OBS. Manually running ("Run macro") on "Use variable" was not changing anything. Manually running the macro on "Set Variable" DOES trigger the change. However, it wouldn't trigger on its own at the source file change.
Current workaround has been to make a new Macro, saying specifically to Run "Set Variables" if the source file content has changed. _As it is now it works._ I'm just curious to understand why the condition on "Use Variable" isn't triggering, nor performing the change on its own.

Thank you so much for this! :D

**Comment by Dexter092** [comment]
Nevermind for the math formula! Within the variables I found what I needed and I followed the same principles you mentioned to make it work. Thank you!
